File: company.py
# 1
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
# Importing the dataset
comp = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Company_Data.csv")
# Checking for null values
comp.isnull().sum()
comp.columns
# Converting into binary
lb = LabelEncoder()
comp["Sales"] = lb.fit_transform(comp["Sales"])
comp["CompPrice"] = lb.fit_transform(comp["CompPrice"])
comp["Income"] = lb.fit_transform(comp["Income"])
comp["Advertising"] = lb.fit_transform(comp["Advertising"])
comp["Population"] = lb.fit_transform(comp["Population"])
comp["Price"] = lb.fit_transform(comp["Price"])
comp["ShelveLoc"] = lb.fit_transform(comp["ShelveLoc"])
comp["Age"] = lb.fit_transform(comp["Age"])
comp["Education"] = lb.fit_transform(comp["Education"])
comp["Urban"] = lb.fit_transform(comp["Urban"])
comp["US"] = lb.fit_transform(comp["US"])
comp['Sales'].unique()
comp['Sales'].value_counts()
colnames = list(comp.columns)
predictors = colnames[:11]
target = colnames[0]
# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(comp, test_size = 0.3)
from sklearn.tree import DecisionTreeClassifier as DT
model = DT(criterion = 'entropy')
model.fit(train[predictors], train[target])
# Prediction on Test Data
pred = model.predict(test[predictors])
pd.crosstab(test[target], pred, rownames=['Actual'], colnames=['Predictions'])
np.mean(pred == test[target]) # Test Data Accuracy 
# Prediction on Train Data
pred = model.predict(train[predictors])
pd.crosstab(train[target], pred, rownames = ['Actual'], colnames = ['Predictions'])
np.mean(pred == train[target]) # Train Data Accuracy

File: company.R
# 1
library(readr)
# Importing the Data
comp <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Company_Data.csv')
str(comp)
# Splitting the data into train and test
library(caTools)
set.seed(0)
split <- sample.split(comp$Sales, SplitRatio = 0.8)
train <- subset(comp, split == TRUE)
test <- subset(comp, split == FALSE)
library(rpart)
model <- rpart(train$Sales ~ ., data = train,control = rpart.control(cp = 0, maxdepth = 3))
# Plot Decision Tree
install.packages("rpart.plot")
library(rpart.plot)
rpart.plot(model, box.palette = "auto", digits = -3)
# Measure the RMSE on Test data
testpred <- predict(model, newdata = test, type = "vector")
# RMSE
accuracy1 <- sqrt(mean(test$Sales - testpred)^2)
accuracy1
# Measure the RMSE on Train data
trainpred <- predict(model, newdata = train, type = "vector")
# RMSE
accuracy2 <- sqrt(mean(train$Sales - trainpred)^2)
accuracy2
# Prune the Decision Tree
# Grow the full tree
fullmodel <- rpart(train$Sales ~ ., data = train,control = rpart.control(cp = 0))
rpart.plot(fullmodel, box.palette = "auto", digits = -3)
# Examine the complexity plot
# Tuning parameter check the value of cp which is giving us minimum cross validation error (xerror)
printcp(fullmodel)   
plotcp(model)
mincp <- model$cptable[which.min(model$cptable[, "xerror"]), "CP"]
# Prune the model based on the optimal cp value
model_pruned_1 <- prune(fullmodel, cp = mincp)
rpart.plot(model_pruned_1, box.palette = "auto", digits = -3)
model_pruned_2 <- prune(fullmodel, cp = 0.02)
rpart.plot(model_pruned_2, box.palette = "auto", digits = -3)
# Measure the RMSE using Full tree
test_pred_fultree <- predict(fullmodel, newdata = test, type = "vector")
# RMSE
accuracy_f <- sqrt(mean(test$Sales - test_pred_fultree)^2)
accuracy_f
# Measure the RMSE using Prune tree - model1
test_pred_prune1 <- predict(model_pruned_1, newdata = test, type = "vector")
# RMSE
accuracy_prune1 <- sqrt(mean(test$Sales - test_pred_prune1)^2)
accuracy_prune1
# Measure the RMSE using Prune tree - model2
test_pred_prune2 <- predict(model_pruned_2, newdata = test, type = "vector")
# RMSE
accuracy_prune2 <- sqrt(mean(test$Sales - test_pred_prune2)^2)
accuracy_prune2
# Prediction for trained data result
train_pred_fultree <- predict(fullmodel,train, type = 'vector')
# RMSE on Train Data
train_accuracy_fultree <- sqrt(mean(train$Sales - train_pred_fultree)^2)
train_accuracy_fultree
# Prediction for trained data result
train_pred_prune1 <- predict(model_pruned_1, train, type = 'vector')
# RMSE on Train Data
train_accuracy_fultree2 <- sqrt(mean(train$Sales - train_pred_prune1)^2)
train_accuracy_fultree2
# Prediction for trained data result
train_pred_prune2 <- predict(model_pruned_2,train, type = 'vector')
# RMSE on Train Data
train_accuracy_fultree2 <- sqrt(mean(train$Sales - train_pred_prune2)^2)
train_accuracy_fultree2

